frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -881,20 +885,22 @@ pub mod pallet {
 
 			// ensure solution claims is better.
 			let mut signed_submissions = Self::signed_submissions();
-			let ejected_a_solution = signed_submissions.len()
-				== T::SignedMaxSubmissions::get().saturated_into::<usize>();
-			let index = Self::insert_submission(&who, &mut signed_submissions, solution, size)
-				.ok_or(Error::<T>::SignedQueueFull)?;
-
-			// collect deposit. Thereafter, the function cannot fail.
-			let deposit = signed_submissions
-				.get(index)
-				.map(|s| s.deposit)
-				.ok_or(Error::<T>::InvalidSubmissionIndex)?;
-			T::Currency::reserve(&who, deposit).map_err(|_| Error::<T>::SignedCannotPayDeposit)?;
-
-			// store the new signed submission.
-			<SignedSubmissions<T>>::put(signed_submissions);
+
+			let (maybe_deposit, ejected_a_solution) =
+				Self::insert_submission(&who, &mut signed_submissions, solution, size);
```

Comment:
I write like 5 different reviews here about how `insert_submission` is quite weird with regard to "check first, write last", as it clearly touches storage (like `unreserve`) but then there are errors after it. But then when you analyse each situation, it _seems_ to be correct and only touch the state when the `SignedQueueFull ` is not triggered. 

UPDATE: well, if the new solution is ejecting an old one, but it does not have enough deposit, then we unreserve the old one, but bail at reserving the new one. This is clearly a BIG issue, and would be good if you write a unit test for it as well (or make sure already unit tests catch it -- since they don't and that's also not a goog sign).

I recommend you to just throw my template of `insert_submission` with its return values out and rewrite this piece.

I think the simplest thing is to just dissolve `insert_submission` and paste it here, and make sure no write happens before all checks pass.  With your refactor, `insert_submission` is nicely slim.

Reply:
Very good catch! Added test in https://github.com/paritytech/substrate/pull/8933/commits/a8e0441f922d55950f622cbcbc3fd114624fe24b; fixed in https://github.com/paritytech/substrate/pull/8933/commits/7dfc184e7b72452ee643f1a952719db140e1c0a8.